### DISCUSSION TITLE
[Ask] Trying to parse a response and the proxy just stops

### DISCUSSION THREAD

**Discussion by alexandresgf** [abhinavsingh/proxy.py#1102] [Q&A] [answered]
Hello,

I'm trying to parse the response from `handle_upstream_data` to check its status code. The thing is, when I try to do it, the application just stops to work and nothing happens. Is there anything that I miss?

Here is my code:

```
def handle_upstream_data(self, raw: memoryview) -> None:
        logger.debug("Creating the Response Parser...")
        response = HttpParser(httpParserTypes.RESPONSE_PARSER)
        logger.debug("Parsing the upstream data...")
        response.parse(raw) # It runs forever
        logger.debug(
            "HTTP Response status code: %d",
            int.from_bytes(response.code),
        )
        super().handle_upstream_data(raw)
```
Just the debug logs below...

![Seleção_021](https://user-images.githubusercontent.com/4246277/156863645-e56c1cf1-31a2-4c4b-9ef8-fe7ccae67e2d.png)

**Comment by abhinavsingh** [comment]
@alexandresgf Can you share the contents of `raw`.  Ideally, you must check for `parser.state` before considering that complete request has been received.  But if `.parse` is running into in infinite loop, that smells like a bug.  Can you please share the raw packet.  Thank you!!!

**Comment by alexandresgf** [comment, marked as answer]
Nevermind, I just was using it in a wrong way. As you said, I checked the state and now everything works!

The final code:

```
def handle_upstream_data(self, raw: memoryview) -> None:
    logger.debug("Creating the Response Parser...")
    response = HttpParser(httpParserTypes.RESPONSE_PARSER)
    logger.debug("Parsing the upstream data...")
    response.parse(raw)

    if response.is_complete:
        logger.debug(
            "Response completed with status code %s",
            response.code,
        )

    super().handle_upstream_data(raw)
```

Thanks!
